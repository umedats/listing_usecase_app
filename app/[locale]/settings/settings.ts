/**
 * Object which will store the necessary settings required to make SP API calls.
 * Each field in this object corresponds to a field in the settings page.
 *
 * region: AWS Region for creating infrastructure
 * accountId: AWS AccountId for creating infrastructure
 * clientId: Client Id part of the selling partner application LWA credentials
 * clientSecret: Client Secret part of the selling partner application LWA credentials
 * refreshToken: Refresh Token used to get the access token
 * sellingPartnerId: Selling Partner Id
 * sellingPartnerIdType: A selling partner identifier type
 * marketplaceId: The MarketplaceId to be used during the Listing Process
 */
export type Settings = {
  region: string;
  accountId: string;
  clientId: string;
  clientSecret: string;
  refreshToken: string;
  sellingPartnerId: string;
  sellingPartnerIdType: string;
  marketplaceId: string;
};

/**
 * returns true if the settings are incomplete.
 * @param settings
 */
export function areSettingsIncomplete(settings: Settings) {
  return (
    !settings.region ||
    !settings.accountId ||
    !settings.clientId ||
    !settings.clientSecret ||
    !settings.refreshToken ||
    !settings.sellingPartnerId ||
    !settings.sellingPartnerIdType ||
    !settings.marketplaceId
  );
}

export const SELLING_PARTNER_TYPE_KEY_VENDOR_CODE = "Vendor Code";
export const SELLING_PARTNER_TYPE_KEY_MERCHANT_ACCOUNT_ID =
  "Merchant Account Id";
