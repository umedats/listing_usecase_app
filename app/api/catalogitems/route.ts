import { NextResponse } from "next/server";
import { headers } from "next/dist/client/components/headers";
import getSPAPIEndpoint, {
  buildCatalogItemsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import {
  IDENTIFIER_TYPES_HEADER,
  IDENTIFIERS_HEADER,
  KEYWORDS_HEADER,
  SEARCH_CATALOG_ITEMS_API_DOC_LINK,
  SEARCH_CATALOG_ITEMS_API_NAME,
  LOCALE_HEADER,
} from "@/app/constants/global";
import nextResponse from "@/app/utils/next-response-factory";
import { Item, ItemSearchResults } from "@/app/sdk/catalogItems_2022-04-01";
import { ProductSearchResult } from "@/app/[locale]/create-offer/product-result-type";
import { Settings } from "@/app/[locale]/settings/settings";
import { SPAPIRequestResponse } from "@/app/model/types";
import { serializeToJsonString } from "@/app/utils/serialization";
import {
  handleSPAPIInvocationError,
  retrieveSettingsAndInvokeAPILogic,
} from "@/app/utils/api";

/**
 * This export forces the NextJs not to pre-render this api route.
 */
export const dynamic = "force-dynamic";

/**
 * Retrieves the list of Catalog Items based on identifiers/keywords by calling the SearchCatalogItems SP-API.
 * @constructor
 */
export async function GET() {
  return retrieveSettingsAndInvokeAPILogic(getCatalogItemsHandler);
}

async function getCatalogItemsHandler(settings: Settings) {
  const { identifiersType, identifiers, keywords, locale } =
    retrieveRequestParamsFromHeader();
  const optionalParams = constructOptionalParametersForApi(
    settings,
    identifiersType,
    identifiers,
    keywords,
    locale,
  );
  const nextResponseOrSPAPIRequestResponse = await fetchCatalogItems(
    settings,
    optionalParams,
  );

  if (nextResponseOrSPAPIRequestResponse instanceof NextResponse) {
    return nextResponseOrSPAPIRequestResponse;
  }

  return fetchItemSearchResults(nextResponseOrSPAPIRequestResponse);
}

function retrieveRequestParamsFromHeader() {
  const requestHeaders = headers();
  const identifiersType = requestHeaders.get(IDENTIFIER_TYPES_HEADER);
  const identifierData = requestHeaders.get(IDENTIFIERS_HEADER);
  const identifiers = identifierData ? identifierData.split(",") : null;
  const keywordsData = requestHeaders.get(KEYWORDS_HEADER);
  const keywords = keywordsData ? keywordsData.split(",") : null;
  const locale = requestHeaders.get(LOCALE_HEADER);

  return { identifiersType, identifiers, keywords, locale };
}

function constructOptionalParametersForApi(
  settings: Settings,
  identifiersType: string | null,
  identifiers: string[] | null,
  keywords: string[] | null,
  locale: string | null,
) {
  const includedData = ["productTypes", "summaries"];

  if (identifiers) {
    const optionalParametersForIdentifiers = {
      identifiers: identifiers,
      identifiersType: identifiersType,
      includedData: includedData,
      locale: locale,
    };

    // Seller ID is mandatory when identifier Type is SKU
    if (identifiersType === "SKU") {
      return {
        sellerId: settings.sellingPartnerId,
        ...optionalParametersForIdentifiers,
      };
    }

    // For non-SKU Identifier Type use case.
    return optionalParametersForIdentifiers;
  }

  if (keywords) {
    return {
      keywords: keywords,
      includedData: includedData,
      locale: locale,
    };
  }
}

async function fetchCatalogItems(settings: Settings, opts: any) {
  // Build Catalog Items API Client.
  const catalogApi = await buildCatalogItemsAPIClient(
    getSPAPIEndpoint(settings.region),
  );

  const reqResponse = {
    apiName: SEARCH_CATALOG_ITEMS_API_NAME,
    apiDocumentationLink: SEARCH_CATALOG_ITEMS_API_DOC_LINK,
    request: {
      marketplaceIds: [settings.marketplaceId],
      ...opts,
    },
    response: {},
  };
  try {
    reqResponse.response = await catalogApi.searchCatalogItems(
      reqResponse.request.marketplaceIds,
      opts,
    );
    return reqResponse;
  } catch (error: any) {
    return handleSPAPIInvocationError(reqResponse, error);
  }
}

function deriveTitle(item: Item) {
  return item.summaries?.[0]?.itemName;
}

function derivePTD(item: Item) {
  return item.productTypes?.[0]?.productType;
}

function fetchItemSearchResults(reqResponse: SPAPIRequestResponse) {
  const searchResults: ItemSearchResults =
    ItemSearchResults.constructFromObject(reqResponse.response, undefined);

  if (searchResults.numberOfResults && searchResults.numberOfResults > 0) {
    // @ts-ignore
    const items: Item[] = searchResults.items;
    const results: ProductSearchResult[] = items.map((item, index) => {
      const searchResult: ProductSearchResult = {
        title: deriveTitle(item),
        asin: item.asin,
        productType: derivePTD(item),
      };
      return searchResult;
    });
    return nextResponse(
      200,
      "OK",
      serializeToJsonString({
        data: results,
        debugContext: [reqResponse],
      }),
    );
  }

  return nextResponse(
    200,
    "OK",
    serializeToJsonString({
      data: [],
      debugContext: [reqResponse],
    }),
  );
}
