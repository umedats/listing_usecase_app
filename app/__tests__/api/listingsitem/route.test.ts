import "isomorphic-fetch";
import { DELETE, GET, POST } from "@/app/api/listingsitem/route";
import { mock, instance, when, reset } from "ts-mockito";
import { getSettings } from "@/app/api/settings/wrapper";
import getSPAPIEndpoint, {
  buildListingsItemsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import { headers } from "next/dist/client/components/headers";
import { ReadonlyHeaders } from "next/dist/server/web/spec-extension/adapters/headers";
import fetch from "jest-fetch-mock";
import { afterEach } from "@jest/globals";
import {
  CREATE_OFFER_USE_CASE,
  LOCALE_HEADER,
  PRODUCT_TYPE_HEADER,
  SKU_HEADER,
  USE_CASE_HEADER,
} from "@/app/constants/global";
import { MOCK_SETTINGS } from "@/app/test-utils/mock-settings";
import {
  mockRejectFetchResponse,
  mockResolveFetchResponse,
} from "@/app/test-utils/mock-fetch";
import { verifyStatusAndBody } from "@/app/test-utils/next-response-helper";
import { ListingsApi } from "@/app/sdk/listingsItems_2021-08-01";
import {
  MOCK_DELETE_LISTINGS_ITEM_API_SUCCESS_RESPONSE,
  MOCK_GET_LISTINGS_ITEM_API_EMPTY_SUMMARY_RESPONSE,
  MOCK_GET_LISTINGS_ITEM_API_SUCCESS_RESPONSE,
  MOCK_PUT_AND_PATCH_LISTINGS_ITEM_API_SUCCESS_RESPONSE,
} from "@/app/test-utils/mock-listings-item-result";
import { NextRequest } from "next/server";
import { serializeToJsonString } from "@/app/utils/serialization";

const SKU = "sku";
const PRODUCT_TYPE = "productType";
const USE_CASE = CREATE_OFFER_USE_CASE;
const US_LOCALE = "en_US";
const MOCK_SPI_API_URL: string = "https://sellingpartnerapi-na.amazon.com";

// @ts-ignore
const mockReadOnlyHeaders: ReadonlyHeaders = mock(ReadonlyHeaders);

const mockedGetListing = jest.fn();
const mockedDeleteListing = jest.fn();
const mockedPutListing = jest.fn();
const mockedPatchListing = jest.fn();

const listingsApi: ListingsApi = new ListingsApi(undefined);
listingsApi.getListingsItem = mockedGetListing;
listingsApi.deleteListingsItem = mockedDeleteListing;
listingsApi.putListingsItem = mockedPutListing;
listingsApi.patchListingsItem = mockedPatchListing;

jest.mock("@/app/sdk/factory/ListingsSPAPIClientsFactory");
const mockedBuildListingsApiClient = jest.mocked(buildListingsItemsAPIClient);
mockedBuildListingsApiClient.mockResolvedValue(listingsApi);
const mockedGetSPAPIEndpoint = jest.mocked(getSPAPIEndpoint);

jest.mock("@/app/api/settings/wrapper");
const mockedGetSettings = jest.mocked(getSettings);

jest.mock("next/dist/client/components/headers");
const mockedHeaders = jest.mocked(headers);

function mockSettings(status: number) {
  mockedGetSettings.mockResolvedValue({
    status: status,
    statusText: "Some Status Text",
    settings: MOCK_SETTINGS,
  });
}

function mockHeaders(
  sku?: string,
  locale?: string,
  productType?: string,
  useCase?: string,
) {
  sku && when(mockReadOnlyHeaders.get(SKU_HEADER)).thenReturn(sku);
  locale && when(mockReadOnlyHeaders.get(LOCALE_HEADER)).thenReturn(locale);
  productType &&
    when(mockReadOnlyHeaders.get(PRODUCT_TYPE_HEADER)).thenReturn(productType);
  useCase && when(mockReadOnlyHeaders.get(USE_CASE_HEADER)).thenReturn(useCase);
  mockedHeaders.mockReturnValue(instance(mockReadOnlyHeaders));
}

function mockGetListingsItem(error?: object) {
  mockedGetSPAPIEndpoint.mockReturnValue(MOCK_SPI_API_URL);
  if (error) {
    mockedGetListing.mockRejectedValue(error);
  } else {
    mockedGetListing.mockResolvedValue(
      MOCK_GET_LISTINGS_ITEM_API_SUCCESS_RESPONSE,
    );
  }
}

function mockDeleteListingsItem(error?: object) {
  mockedGetSPAPIEndpoint.mockReturnValue(MOCK_SPI_API_URL);
  if (error) {
    mockedDeleteListing.mockRejectedValue(error);
  } else {
    mockedDeleteListing.mockResolvedValue(
      MOCK_DELETE_LISTINGS_ITEM_API_SUCCESS_RESPONSE,
    );
  }
}

function mockPutListingsItem(error?: object) {
  mockedGetSPAPIEndpoint.mockReturnValue(MOCK_SPI_API_URL);
  if (error) {
    mockedPutListing.mockRejectedValue(error);
  } else {
    mockedPutListing.mockResolvedValue(
      MOCK_PUT_AND_PATCH_LISTINGS_ITEM_API_SUCCESS_RESPONSE,
    );
  }
}

function mockPatchListingsItem(error?: object) {
  mockedGetSPAPIEndpoint.mockReturnValue(MOCK_SPI_API_URL);
  if (error) {
    mockedPatchListing.mockRejectedValue(error);
  } else {
    mockedPatchListing.mockResolvedValue(
      MOCK_PUT_AND_PATCH_LISTINGS_ITEM_API_SUCCESS_RESPONSE,
    );
  }
}

function buildPostNextRequest() {
  return new NextRequest("http://localhost:3000/api/listingsitem", {
    method: "POST",
    headers: {
      sku: SKU,
      locale: US_LOCALE,
      productType: PRODUCT_TYPE,
      useCase: USE_CASE,
    },
    body: serializeToJsonString({
      initialListing: {
        externally_assigned_product_identifier: [
          {
            type: "upc",
            value: "650450120249",
            marketplace_id: "ATVPDKIKX0DER",
          },
        ],
      },
      currentListing: {
        externally_assigned_product_identifier: [
          {
            type: "upc",
            value: "650450120250",
            marketplace_id: "ATVPDKIKX0DER",
          },
        ],
      },
    }),
  });
}

function buildDeleteNextRequest() {
  return new NextRequest("http://localhost:3000/api/listingsitem", {
    method: "DELETE",
    headers: {
      sku: SKU,
      locale: US_LOCALE,
    },
  });
}

describe("Test PUT ListingsItem SP-API", () => {
  beforeEach(() => {
    fetch.resetMocks();
    jest.clearAllMocks();
  });

  afterEach(() => {
    reset(mockReadOnlyHeaders);
  });

  test("returns 200 on successful call to Put Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(SKU, US_LOCALE, PRODUCT_TYPE, USE_CASE);
    mockPutListingsItem();
    mockResolveFetchResponse(
      200,
      MOCK_PUT_AND_PATCH_LISTINGS_ITEM_API_SUCCESS_RESPONSE,
    );

    await verifyStatusAndBody(await POST(buildPostNextRequest()), 200, true);
  });

  test("returns 500 on exception from settings response when calling Put Listings Item API", async () => {
    mockSettings(500);
    mockHeaders(SKU, US_LOCALE, PRODUCT_TYPE, USE_CASE);
    mockPutListingsItem();
    mockRejectFetchResponse(500);

    await verifyStatusAndBody(await POST(buildPostNextRequest()), 500, false);
  });

  test("returns 400 exception from missing SKU header from Put Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(undefined, US_LOCALE, PRODUCT_TYPE, USE_CASE);
    mockPutListingsItem({
      status: 400,
    });
    mockRejectFetchResponse(400);

    await verifyStatusAndBody(await POST(buildPostNextRequest()), 400, false);
  });

  test("returns 400 exception from missing locale header from Put Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(SKU, undefined, PRODUCT_TYPE, USE_CASE);
    mockPutListingsItem({
      status: 400,
    });
    mockRejectFetchResponse(400);

    await verifyStatusAndBody(await POST(buildPostNextRequest()), 400, false);
  });

  test("returns 400 exception from missing Product Type header from Put Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(SKU, US_LOCALE, undefined, USE_CASE);
    mockPutListingsItem({
      status: 400,
    });
    mockRejectFetchResponse(400);

    await verifyStatusAndBody(await POST(buildPostNextRequest()), 400, false);
  });

  test("returns 400 exception from missing Use case header from Put Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(SKU, US_LOCALE, PRODUCT_TYPE, undefined);
    mockPutListingsItem({
      status: 400,
    });
    mockRejectFetchResponse(400);

    await verifyStatusAndBody(await POST(buildPostNextRequest()), 400, false);
  });

  test("returns 400 exception from missing current Listing data in request body", async () => {
    mockSettings(200);
    mockHeaders(SKU, US_LOCALE, PRODUCT_TYPE, USE_CASE);
    mockPutListingsItem({
      status: 400,
    });
    mockRejectFetchResponse(400);
    const nextRequest = new NextRequest(
      "http://localhost:3000/api/submitListing",
      {
        method: "POST",
        headers: {
          sku: SKU,
          locale: US_LOCALE,
          productType: PRODUCT_TYPE,
          useCase: USE_CASE,
        },
        body: serializeToJsonString({
          initialListing: {},
        }),
      },
    );

    await verifyStatusAndBody(await POST(nextRequest), 400, false);
  });

  test("returns 400 exception from missing initial Listing data in request body", async () => {
    mockSettings(200);
    mockHeaders(SKU, US_LOCALE, PRODUCT_TYPE, USE_CASE);
    mockPutListingsItem({
      status: 400,
    });
    mockRejectFetchResponse(400);
    const nextRequest = new NextRequest(
      "http://localhost:3000/api/listingsitem",
      {
        method: "POST",
        headers: {
          sku: SKU,
          locale: US_LOCALE,
          productType: PRODUCT_TYPE,
          useCase: USE_CASE,
        },
        body: serializeToJsonString({
          currentListing: {},
        }),
      },
    );

    await verifyStatusAndBody(await POST(nextRequest), 400, false);
  });

  test("returns 500 exception from invoking Put Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(SKU, US_LOCALE, PRODUCT_TYPE, USE_CASE);
    mockPutListingsItem({
      status: 500,
    });
    mockRejectFetchResponse(500);

    await verifyStatusAndBody(await POST(buildPostNextRequest()), 500, false);
  });
});

describe("Test GET ListingsItem SP-API", () => {
  beforeEach(() => {
    fetch.resetMocks();
    jest.clearAllMocks();
  });

  afterEach(() => {
    reset(mockReadOnlyHeaders);
  });

  test("returns 200 on successful call to Get Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(SKU, US_LOCALE);
    mockGetListingsItem();
    mockResolveFetchResponse(200, MOCK_GET_LISTINGS_ITEM_API_SUCCESS_RESPONSE);

    await verifyStatusAndBody(await GET(), 200, true);
  });

  test("returns 200 on successful call to Get Listings Item API with no summary data", async () => {
    mockSettings(200);
    mockHeaders(SKU, US_LOCALE);
    mockedGetSPAPIEndpoint.mockReturnValue(MOCK_SPI_API_URL);
    mockedGetListing.mockResolvedValue(
      MOCK_GET_LISTINGS_ITEM_API_EMPTY_SUMMARY_RESPONSE,
    );
    mockResolveFetchResponse(
      200,
      MOCK_GET_LISTINGS_ITEM_API_EMPTY_SUMMARY_RESPONSE,
    );

    await verifyStatusAndBody(await GET(), 200, true);
  });

  test("returns 200 on 404 response from Get Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(SKU, US_LOCALE);
    mockGetListingsItem({ status: 404 });
    mockResolveFetchResponse(200, {});

    await verifyStatusAndBody(await GET(), 200, true);
  });

  test("returns 500 on exception from settings response when calling Get Listings Item API", async () => {
    mockSettings(500);
    mockHeaders(SKU, US_LOCALE);
    mockGetListingsItem();
    mockRejectFetchResponse(500);

    await verifyStatusAndBody(await GET(), 500, false);
  });

  test("returns 400 exception from missing SKU header from Get Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(undefined, US_LOCALE);
    mockGetListingsItem({ status: 400 });
    mockRejectFetchResponse(400);

    await verifyStatusAndBody(await GET(), 400, false);
  });

  test("returns 400 on call to Get Listings Item API on missing locale", async () => {
    mockSettings(200);
    mockHeaders(SKU, undefined);
    mockGetListingsItem();
    mockRejectFetchResponse(400);

    await verifyStatusAndBody(await GET(), 400, false);
  });

  test("returns 500 exception from invoking Get Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(SKU, US_LOCALE);
    mockGetListingsItem({ status: 500 });
    mockRejectFetchResponse(500);

    await verifyStatusAndBody(await GET(), 500, false);
  });
});

describe("Test DELETE ListingsItem API", () => {
  beforeEach(() => {
    fetch.resetMocks();
    jest.clearAllMocks();
  });

  afterEach(() => {
    reset(mockReadOnlyHeaders);
  });

  test("returns 200 on successful call to DELETE Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(SKU, US_LOCALE);
    mockDeleteListingsItem();
    mockResolveFetchResponse(
      200,
      MOCK_DELETE_LISTINGS_ITEM_API_SUCCESS_RESPONSE,
    );

    await verifyStatusAndBody(
      await DELETE(buildDeleteNextRequest()),
      200,
      true,
    );
  });

  test("returns 500 on exception from settings response when calling DELETE Listings Item API", async () => {
    mockSettings(500);
    mockHeaders(SKU, US_LOCALE);
    mockDeleteListingsItem();
    mockRejectFetchResponse(500);

    await verifyStatusAndBody(
      await DELETE(buildDeleteNextRequest()),
      500,
      false,
    );
  });

  test("returns 400 exception from missing SKU header from Delete Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(undefined, US_LOCALE);
    mockDeleteListingsItem({
      status: 400,
    });
    mockRejectFetchResponse(400);

    await verifyStatusAndBody(
      await DELETE(buildDeleteNextRequest()),
      400,
      false,
    );
  });

  test("returns 400 exception from missing locale header from Delete Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(SKU, undefined);
    mockDeleteListingsItem({
      status: 400,
    });
    mockRejectFetchResponse(400);

    await verifyStatusAndBody(
      await DELETE(buildDeleteNextRequest()),
      400,
      false,
    );
  });

  test("returns 500 exception from invoking Delete Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(SKU, US_LOCALE);
    mockDeleteListingsItem({
      status: 500,
    });
    mockRejectFetchResponse(500);
    await verifyStatusAndBody(
      await DELETE(buildDeleteNextRequest()),
      500,
      true,
    );
  });
});

describe("Test PATCH ListingsItem API", () => {
  beforeEach(() => {
    fetch.resetMocks();
    jest.clearAllMocks();
  });

  afterEach(() => {
    reset(mockReadOnlyHeaders);
  });

  test("returns 200 on successful call to Patch Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(SKU, US_LOCALE, PRODUCT_TYPE, "UpdateListing");
    mockPatchListingsItem();
    mockResolveFetchResponse(
      200,
      MOCK_PUT_AND_PATCH_LISTINGS_ITEM_API_SUCCESS_RESPONSE,
    );

    await verifyStatusAndBody(await POST(buildPostNextRequest()), 200, true);
  });

  test("returns 500 exception from invoking Patch Listings Item API", async () => {
    mockSettings(200);
    mockHeaders(SKU, US_LOCALE, PRODUCT_TYPE, "UpdateListing");
    mockPatchListingsItem({
      status: 500,
    });
    mockRejectFetchResponse(500);

    await verifyStatusAndBody(await POST(buildPostNextRequest()), 500, true);
  });
});
