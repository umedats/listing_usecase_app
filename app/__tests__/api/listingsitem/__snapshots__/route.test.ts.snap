// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test DELETE ListingsItem API returns 200 on successful call to DELETE Listings Item API 1`] = `
{
  "data": {
    "sku": "My SKU",
    "status": "ACCEPTED",
    "submissionId": "Some UUID",
  },
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/listings-items-api-v2021-08-01-reference#delete-listings2021-08-01itemsselleridsku",
      "apiName": "DeleteListingsItem",
      "request": {
        "issueLocale": "en_US",
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
        "sellerId": "ABCD878787R",
        "sku": "sku",
      },
      "response": {
        "sku": "My SKU",
        "status": "ACCEPTED",
        "submissionId": "Some UUID",
      },
    },
  ],
}
`;

exports[`Test DELETE ListingsItem API returns 500 exception from invoking Delete Listings Item API 1`] = `
{
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/listings-items-api-v2021-08-01-reference#delete-listings2021-08-01itemsselleridsku",
      "apiName": "DeleteListingsItem",
      "request": {
        "issueLocale": "en_US",
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
        "sellerId": "ABCD878787R",
        "sku": "sku",
      },
      "response": {
        "status": 500,
      },
    },
  ],
}
`;

exports[`Test GET ListingsItem SP-API returns 200 on 404 response from Get Listings Item API 1`] = `
{
  "data": {},
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/listings-items-api-v2021-08-01-reference#get-listings2021-08-01itemsselleridsku",
      "apiName": "GetListingsItem",
      "request": {
        "includedData": [
          "attributes",
          "summaries",
          "issues",
        ],
        "issueLocale": "en_US",
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
        "sellerId": "ABCD878787R",
        "sku": "sku",
      },
      "response": {
        "status": 404,
      },
    },
  ],
}
`;

exports[`Test GET ListingsItem SP-API returns 200 on successful call to Get Listings Item API 1`] = `
{
  "data": {
    "attributes": {
      "color": [
        {
          "language_tag": "en_US",
          "marketplace_id": "ATVPDKIKX0DER",
          "value": "Two-Tone",
        },
      ],
    },
    "productType": "SUNGLASSES",
  },
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/listings-items-api-v2021-08-01-reference#get-listings2021-08-01itemsselleridsku",
      "apiName": "GetListingsItem",
      "request": {
        "includedData": [
          "attributes",
          "summaries",
          "issues",
        ],
        "issueLocale": "en_US",
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
        "sellerId": "ABCD878787R",
        "sku": "sku",
      },
      "response": {
        "attributes": {
          "color": [
            {
              "language_tag": "en_US",
              "marketplace_id": "ATVPDKIKX0DER",
              "value": "Two-Tone",
            },
          ],
        },
        "sku": "S014-A03",
        "summaries": [
          {
            "asin": "B003VDOHBG",
            "conditionType": "new_new",
            "createdDate": "2023-02-10T01:29:04.021Z",
            "itemName": "SKAGEN DENMARK Unisex Sunglasses Steel Aviator Frame #S014-A03",
            "lastUpdatedDate": "2023-02-10T01:29:09.595Z",
            "mainImage": {
              "height": 392,
              "link": "https://m.media-amazon.com/images/I/41cVJ8Gz6TL.jpg",
              "width": 444,
            },
            "marketplaceId": "ATVPDKIKX0DER",
            "productType": "SUNGLASSES",
            "status": [],
          },
        ],
      },
    },
  ],
}
`;

exports[`Test GET ListingsItem SP-API returns 200 on successful call to Get Listings Item API with no summary data 1`] = `
{
  "data": {
    "attributes": {},
  },
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/listings-items-api-v2021-08-01-reference#get-listings2021-08-01itemsselleridsku",
      "apiName": "GetListingsItem",
      "request": {
        "includedData": [
          "attributes",
          "summaries",
          "issues",
        ],
        "issueLocale": "en_US",
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
        "sellerId": "ABCD878787R",
        "sku": "sku",
      },
      "response": {
        "attributes": {},
        "sku": "S014-A03",
        "summaries": [],
      },
    },
  ],
}
`;

exports[`Test PATCH ListingsItem API returns 200 on successful call to Patch Listings Item API 1`] = `
{
  "data": {
    "sku": "My SKU",
    "status": "ACCEPTED",
    "submissionId": "Some UUID",
  },
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/listings-items-api-v2021-08-01-reference#patch-listings2021-08-01itemsselleridsku",
      "apiName": "PatchListingsItem",
      "request": {
        "body": {
          "patches": [
            {
              "op": "replace",
              "path": "/attributes/externally_assigned_product_identifier",
              "value": [
                {
                  "marketplace_id": "ATVPDKIKX0DER",
                  "type": "upc",
                  "value": "650450120250",
                },
              ],
            },
          ],
          "productType": "productType",
        },
        "issueLocale": "en_US",
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
        "sellerId": "ABCD878787R",
        "sku": "sku",
      },
      "response": {
        "sku": "My SKU",
        "status": "ACCEPTED",
        "submissionId": "Some UUID",
      },
    },
  ],
}
`;

exports[`Test PATCH ListingsItem API returns 500 exception from invoking Patch Listings Item API 1`] = `
{
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/listings-items-api-v2021-08-01-reference#patch-listings2021-08-01itemsselleridsku",
      "apiName": "PatchListingsItem",
      "request": {
        "body": {
          "patches": [
            {
              "op": "replace",
              "path": "/attributes/externally_assigned_product_identifier",
              "value": [
                {
                  "marketplace_id": "ATVPDKIKX0DER",
                  "type": "upc",
                  "value": "650450120250",
                },
              ],
            },
          ],
          "productType": "productType",
        },
        "issueLocale": "en_US",
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
        "sellerId": "ABCD878787R",
        "sku": "sku",
      },
      "response": {
        "status": 500,
      },
    },
  ],
}
`;

exports[`Test PUT ListingsItem SP-API returns 200 on successful call to Put Listings Item API 1`] = `
{
  "data": {
    "sku": "My SKU",
    "status": "ACCEPTED",
    "submissionId": "Some UUID",
  },
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/listings-items-api-v2021-08-01-reference#put-listings2021-08-01itemsselleridsku",
      "apiName": "PutListingsItem",
      "request": {
        "body": {
          "attributes": {
            "externally_assigned_product_identifier": [
              {
                "marketplace_id": "ATVPDKIKX0DER",
                "type": "upc",
                "value": "650450120250",
              },
            ],
          },
          "productType": "productType",
          "requirements": "LISTING_OFFER_ONLY",
        },
        "issueLocale": "en_US",
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
        "sellerId": "ABCD878787R",
        "sku": "sku",
      },
      "response": {
        "sku": "My SKU",
        "status": "ACCEPTED",
        "submissionId": "Some UUID",
      },
    },
  ],
}
`;
